examples/server-communication__offline/cypress/integration/offline-spec.js

```diff
@@ -1,177 +1,87 @@
-    cy.visit('/')
-    cy.contains('#network-status', 'online')
-    .wait(1000) // for demo purpose
-
-    goOffline()
-    cy.contains('#network-status', 'offline')
-    .wait(1000) // for demo purpose
-  })
-
-  it('shows error if we stub the network call', () => {
-    assertOnline()
-    cy.visit('/')
-    cy.intercept(`${url}*`, { forceNetworkError: true }).as('users')
-    cy.get('#load-users').click()
-    cy.contains('#users', 'Problem fetching users Failed to fetch')
-
-    // cannot wait for the intercept that forces network error
-    // https://github.com/cypress-io/cypress/issues/9062
-    // cy.wait('@users', { timeout: 1000 }) // the network call happens
-  })
-
-  it('shows error trying to fetch users in offline mode', () => {
-    cy.visit('/')
-    assertOnline()
-
-    // since this call returns a promise, must tell Cypress to wait
-    // for it to be resolved
-    goOffline()
-    assertOffline()
+  context('when online', () => {
+    beforeEach(() => {
+      cy.visit('/', {
+        onBeforeLoad(win) {
+          // use window.navigator.onLine property to
+          // mock browser is offline or online
+          // https://caniuse.com/online-status
+          cy.stub(win.navigator, 'onLine', true).as('online')
+        }
+      })
+      cy.intercept('*').as('users')
 
-    cy.get('#load-users').click()
-    cy.contains('#users', 'Problem fetching users Failed to fetch')
+      cy.wrap(window).its('navigator.onLine').should('be.true')
+    })
 
-    // now let's go back online and fetch the users
-    goOnline()
-    assertOnline()
-    cy.get('#load-users').click()
-    cy.get('.user').should('have.length', 3)
+    it('shows network status as online', () => {
+      cy.contains('#network-status', 'online')
+    })
+  
+    it('fetches users', () => {
+      // let's spy on the "fetch" method the app calls
+      cy.window().then((w) => cy.spy(w, 'fetch').withArgs(`${url}?_limit=3`).as('fetchUsers'))
+      cy.get('#load-users').click()
+      cy.get('.user').should('have.length', 3)
+      cy.get('@fetchUsers').should('have.been.calledOnce')
+    })
   })
 
-  it('makes fetch request when offline', () => {
-    cy.visit('/')
-
-    goOffline()
-    assertOffline()
-
-    // let's spy on the "fetch" method the app calls
-    cy.window().then((w) => cy.spy(w, 'fetch').withArgs(`${url}?_limit=3`).as('fetchUsers'))
-
-    cy.get('#load-users').click()
-    cy.get('@fetchUsers').should('have.been.calledOnce')
-
-    // now let's go back online and fetch the users
-    goOnline()
-    assertOnline()
-    cy.get('#load-users').click()
-    cy.get('.user').should('have.length', 3)
-    cy.get('@fetchUsers').should('have.been.calledTwice')
-  })
+  context('when offline', () => {
+    beforeEach(() => {
+      cy.visit('/', {
+        onBeforeLoad(win) {
+          cy.stub(win.navigator, 'onLine', false).as('online')
+        }
+      })
+      cy.intercept('*', { forceNetworkError: true }).as('users')
```

Comment:
should intercept before `cy.visit`

also, why is this aliased as `users`?

Reply:
Blerg, yeah, cuz you're doing `cy.intercept('*')`...

Just kinda weird cuz this pattern is non-deterministic. There's a race condition between when the page is loading resources after `cy.visit`, and when the `cy.intercept` runs.

This non-determinism will cause the same kind of flake as in https://github.com/jennifer-shehane/cypress-intercept-times-flake if users adopt it in their tests.

So, right now, this test is testing "what happens if I turn off the Internet at some random point after my cy.visit", randomness is usually not a practice we want to encourage.

Reply:
>So, right now, this test is testing "what happens if I turn off the Internet at some random point after my cy.visit?", randomness is usually not a practice we want to encourage.

Yah, it's kind of a weird test, I agree. The original test was testing going offline and online many times within a single test, but I don't think that's possible with cy.intercept() and I'm not sure this is a use case people need to test anyway.

I'm open to suggestions, was trying to replace the CDP test if possible but I can just ditch this effort if it's not a good test.

Reply:
IF a user wants to test having everything offline, I think a solid example test would be to use `cy.intercept()` to disable network traffic, then do an action (`cy.click()`, ...) that triggers a network request, then assert that the failure was handled as expected.

Alternatively, if the user just wants to test that a particular endpoint is offline, they could do `cy.intercept('http://api.example.com/**', { forceNetworkError: true })` followed by a `cy.visit('http://www.example.com/')` and they wouldn't have this problem.